README.md
# PyQuantity

A package to work with physical quantity in Python.

## Global Informations

 * Author: LostPy
 * Version: 0.0.1
 * Last version release: None
 * Documentation
 * Requirements:
   - Python 3.6+

CHANGELOG.md
# Changelog

No changelog.

test.py
from pyquantity import PrefixEnum, BaseUnit, Quantity


def test_prefix():
	print([mb for mb in dir(PrefixEnum) if not mb.startswith('_')])
	print(PrefixEnum.hecto.name, PrefixEnum.hecto.symbol, PrefixEnum.hecto.value)
	
	PrefixEnum.add_prefix('custom', 'R', 1e30)
	print(PrefixEnum.custom)

	PrefixEnum.remove_prefix('custom')
	print([mb for mb in dir(PrefixEnum) if not mb.startswith('_')])


def base_unit():
	meter = BaseUnit.meter()
	print(meter)
	print('meter is a base unit: ', meter.is_base_unit())
	print('meter is a derived unit: ', meter.is_derived_unit())
	print('str(meter) = ', str(meter))


def test_quantity():
	q = Quantity(10.5, BaseUnit.meter())
	print(q)
	print('str(q): ', str(q))
	print('q.hecto(): ', q.hecto())
	print('q.n: ', q.n)
	print('q.G: ', q.G)
	q.k = 2
	print('q.value', q.value)


def test_operations():
	q1 = Quantity(10.3, BaseUnit.meter())
	q2 = Quantity(21.3, BaseUnit.meter())
	print(q1 < q2)
	print(q1 == q2)
	print(q1 > 15)
	print(q2 > 15)

	print(-q1)
	print(+q2)
	print(q1 * 4)
	print(q1 + q2)


if __name__ == '__main__':
	#test_prefix()
	#base_unit()
	#test_quantity()
	test_operations()


pyquantity/quantity.py
from typing import Union
import operator
from copy import deepcopy
from math import log10

from . import Prefix, PrefixEnum
from .unit import Unit
from . import errors


class Quantity:
	_unit = None  # define the units of quantity
	_base_prefix = PrefixEnum.none  # define the base prefix of quantity. For example, for the mass it's 'kilo'.

	def __init__(self, value: float, unit: Union[str, Unit], base_prefix: Union[str, float, Prefix] = PrefixEnum.none, prefix: Prefix = None):
		"""Initialize an instance of Quantity
	
		Parameters
		----------
		None
		"""
		self.base_prefix = base_prefix
		if (prefix is None and self._base_prefix.value == 1) or (isinstance(prefix, Prefix) and prefix.value == 1):
			# value with a prefix of value = 1
			self._value = value

		elif prefix is None or (isinstance(prefix, Prefix) and prefix.value == slef.base_prefix.value):
			# value with the base prefix
			self._value = PrefixEnum.convert_value(value, to_=PrefixEnum.none, from_=self._base_prefix)

		elif isinstance(prefix, Prefix):
			# value with other prefix
			self._value = PrefixEnum.convert_value(value, to_=PrefixEnum.none, from_=prefix)

		else:
			raise ValueError(f"'prefix' argument must be a Prefix, not {type(prefix)}")

		self.unit = unit

	def __repr__(self) -> str:
		"""repr(self)"""
		return f"<{self.name}: {self._value} {self.unit.symbol}>"

	def __str__(self) -> str:
		"""str(self)"""
		exp = int(log10(abs(self._value)))
		if exp > 24:
			return f"{self.Y:.3e} {self.unit.symbol_with_prefix(PrefixEnum.yotta)}"
		elif exp < -24:
			return f"{self.y:.3e} {self.unit.symbol_with_prefix(PrefixEnum.yocto)}"
		rank = exp // 3
		prefix = PrefixEnum.get_prefix_with_rank(rank)
		if abs(rank) > 8:
			return f"{self.value_to_prefix(prefix):.3e} {self.unit.symbol_with_prefix(prefix)}"
		return f"{self.value_to_prefix(prefix):.3f} {self.unit.symbol_with_prefix(prefix)}"

	def __format__(self, fmt_spec: str = "") -> str:
		"""format(self, str)"""
		return format(str(self), format_spec=fmt_spec)

	def __add__(self, other):
		"""self + other"""
		return self.__addition(other, operator.add)

	def __sub__(self, other):
		"""self - other"""
		return self.__addition(other, operator.sub)

	def __mul__(self, other):
		"""self * other"""
		return self.__product(other, operator.mul)

	def __div__(self, other):
		"""self / other"""
		return self.__product(other, operator.div)

	def __pow__(self, other):
		"""self ** other"""
		raise NotImplemented

	def __iadd__(self, other):
		"""self += other"""
		self.__assignement(other, operator.iadd)

	def __isub__(self, other):
		"""self -= other"""
		self.__assignement(other, operator.isub)

	def __pos__(self):
		""" + self"""
		return self.__unary_operation(operator.pos)

	def __neg__(self):
		"""- self"""
		return self.__unary_operation(operator.neg)

	def __abs__(self):
		""" abs(self)"""
		return self.__unary_operation(operator.abs)

	def __int__(self) -> int:
		"""int(self)"""
		return int(self._value)

	def __float__(self) -> float:
		"""float(self)"""
		return float(self._value)

	def __lt__(self, other) -> bool:
		"""self < other"""
		return self.__compare(other, operator.lt)

	def __le__(self, other) -> bool:
		"""self <= other"""
		return self.__compare(other, operator.le)

	def __eq__(self, other) -> bool:
		"""self == other"""
		return self.__compare(other, operator.eq)

	def __ne__(self, other) -> bool:
		"""self != other"""
		return self.__compare(other, operator.ne)

	def __gt__(self, other) -> bool:
		"""self > other"""
		return self.__compare(other, operator.gt)

	def __ge__(self, other) -> bool:
		"""self >= other"""
		return self.__compare(other, operator.ge)

	def __addition(self, other, op):
		"""Method for addition or similar
		other can be a int or float. If other is a int or float it is defined with the base prefix.

		Parameters
		----------
		other
			The second member of operation
		op : operator
			The operator to use

		Returns
		-------
		quantity : Quantity
			A new quantity instance of same type than self
		"""
		new = deepcopy(self)
		if type(other) == type(self):
			new._value = op(self._value, other.value)
		elif isinstance(other, (int, float)):
			new._value = op(self._value, float(other))
		else:
			raise ValueError(f"Can't do a addition between {type(self)} and {type(other)}")
		return new

	def __product(self, other, op):
		"""Method for product or similar
		other can be a int or float.

		Parameters
		----------
		other
			The second member of operation
		op : operator
			The operator to use
		
		Returns
		-------
		quantity : Quantity
			A new quantity instance

		Raises
		------
		NotImplemented
			If other is an quantity
		"""
		new = deepcopy(self)
		if isinstance(other, (int, float)):
			new._value = op(self._value, float(other))
		elif isinstance(other, Quantity):
			raise NotImplemented
		else:
			raise ValueError(f"Can't do a addition between {type(self)} and {type(other)}")
		return new

	def __assignement(self, other, op):
		"""Method for assignment operators.
		other can be a int or float. If other is a int or float, it is defined with the base prefix.

		Parameters
		----------
		other
			The second member of operation
		op : operator
			The operator to use
		"""
		if type(other) == type(self):
			op(self._value, other.value)
		elif isinstance(other, (int, float)):
			op(self._value, PrefixEnum.convert_value(other, to_=PrefixEnum.none, from_=self._base_prefix))
		raise ValueError(f"Can't do a assignment between {type(self)} and {type(other)}")

	def __unary_operation(self, op):
		"""Method for unary operator
		
		Parameters
		----------
		op : operator
			The operator to use
		
		Returns
		-------
		quantity : Quantity
			A new quantity instance of same type than self
		"""
		new = deepcopy(self)
		new._value = op(self._value)
		return new

	def __compare(self, other, op) -> bool:
		"""Method to compare self and other with a operator.
		other can be a int or float. If other is a int or float, it is defined with the base prefix.

		Parameters
		----------
		other
			The second member of operation
		op : operator
			The operator to use
		
		Returns
		-------
		result : bool
			The result of comparison
		"""
		if type(other) == type(self):
			return op(self._value, other.value)
		elif isinstance(other, (int, float)):
			return op(self._value, PrefixEnum.convert_value(other, to_=PrefixEnum.none, from_=self._base_prefix))
		raise ValueError(f"Can't compare '{type(self)}' with '{type(other)}'")

	def _format_real_value(self) -> str:
		if abs(self._value) >= 1000:
			return f"{self._value:.3e}"
		return f"{self._value:.3f}"

	def value_to_prefix(self, prefix: Prefix) -> float:
		if not isinstance(prefix, Prefix):
			raise ValueError(f"'prefix' must be a Prefix, not a {type(prefix)}")
		return PrefixEnum.convert_value(self.value, to_=prefix)

	def set_value_from_prefix(self, new_value: float, prefix: Prefix):
		try:
			new_value = float(new_value)
		except Exception:
			raise ValueError(f"'new_value' must be a float, not a {type(new_value)}")

		if not isinstance(prefix, Prefix):
			raise ValueError(f"'prefix' must be a Prefix, not a {type(prefix)}")

		self._value = PrefixEnum.convert_value(new_value, to_=PrefixEnum.none, from_=prefix)

	def yotta(self) -> float:
		return self.value_to_prefix(PrefixEnum.yotta)

	def zetta(self) -> float:
		return self.value_to_prefix(PrefixEnum.zetta)

	def exa(self) -> float:
		return self.value_to_prefix(PrefixEnum.exa)

	def peta(self) -> float:
		return self.value_to_prefix(PrefixEnum.peta)

	def tera(self) -> float:
		return self.value_to_prefix(PrefixEnum.tera)

	def giga(self) -> float:
		return self.value_to_prefix(PrefixEnum.giga)

	def mega(self) -> float:
		return self.value_to_prefix(PrefixEnum.mega)

	def kilo(self) -> float:
		return self.value_to_prefix(PrefixEnum.kilo)

	def hecto(self) -> float:
		return self.value_to_prefix(PrefixEnum.hecto)

	def deca(self) -> float:
		return self.value_to_prefix(PrefixEnum.deca)

	def no_prefix(self) -> float:
		return self.value

	def deci(self) -> float:
		return self.value_to_prefix(PrefixEnum.deci)

	def centi(self) -> float:
		return self.value_to_prefix(PrefixEnum.centi)

	def mili(self) -> float:
		return self.value_to_prefix(PrefixEnum.mili)

	def micro(self) -> float:
		return self.value_to_prefix(PrefixEnum.micro)

	def nano(self) -> float:
		return self.value_to_prefix(PrefixEnum.nano)

	def pico(self) -> float:
		return self.value_to_prefix(PrefixEnum.pico)

	def femto(self) -> float:
		return self.value_to_prefix(PrefixEnum.femto)

	def atto(self) -> float:
		return self.value_to_prefix(PrefixEnum.atto)

	def zepto(self) -> float:
		return self.value_to_prefix(PrefixEnum.zepto)

	def yocto(self) -> float:
		return self.value_to_prefix(PrefixEnum.yocto)

	def set_yotta(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.yotta)

	def set_zetta(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.zetta)

	def set_exa(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.exa)

	def set_peta(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.peta)

	def set_tera(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.tera)

	def set_giga(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.giga)

	def set_mega(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.mega)

	def set_kilo(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.kilo)

	def set_hecto(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.hecto)

	def set_deca(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.deca)

	def set_no_prefix(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.none)

	def set_deci(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.deci)

	def set_centi(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.centi)

	def set_mili(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.mili)

	def set_micro(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.micro)

	def set_nano(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.nano)

	def set_pico(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.pico)

	def set_femto(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.femto)

	def set_atto(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.atto)

	def set_zepto(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.zepto)

	def set_yocto(self, new_value: float):
		self.set_value_from_prefix(new_value, PrefixEnum.yocto)

	@classmethod
	def get_name(cls) -> str:
		return cls.__name__

	@property
	def unit(self) -> Unit:
		return self._unit

	@unit.setter
	def unit(self, new_unit: Union[str, Unit]):
		if isinstance(new_unit, str):
			self._unit = Unit(new_unit, new_unit)
		elif isinstance(new_unit, Unit):
			self._unit = new_unit
		else:
			raise ValueError(f"'unit' must be a string which define the symbol of unit or an instance of Unit, not {type(new_unit)}")

	@property
	def base_prefix(self) -> Prefix:
		return self._base_prefix

	@base_prefix.setter
	def base_prefix(self, prefix: Union[str, float, Prefix]):
		if isinstance(prefix, Prefix):
			self._base_prefix = prefix
		elif isinstance(prefix, str) or isinstance(prefix, float):
			PrefixEnum.get_prefix(prefix)
		else:
			raise ValueError(f"'prefix' must be a str, a float or a Prefix, not a {type(prefix)}")

	@property
	def value(self) -> float:
		return self._value

	@value.setter
	def value(self, new_value: float):
		if isinstance(new_value, float) and self._base_prefix.value == 1:
			self._value = new_value
		elif isinstance(new_value, float):
			self.set_value_from_prefix(new_value, to_=PrefixEnum.none, from_=self._base_prefix)
		raise ValueError(f"'value' must be a float, not {type(new_value)}")

	@property
	def name(self) -> str:
		return self.get_name()

	@property
	def Y(self) -> float:
		return self.yotta()

	@property
	def Z(self) -> float:
		return self.zetta()

	@property
	def E(self) -> float:
		return self.exa()

	@property
	def P(self) -> float:
		return self.peta()

	@property
	def T(self) -> float:
		return self.tera()

	@property
	def G(self) -> float:
		return self.giga()

	@property
	def M(self) -> float:
		return self.mega()

	@property
	def k(self) -> float:
		return self.kilo()

	@property
	def h(self) -> float:
		return self.hecto()

	@property
	def da(self) -> float:
		return self.deca()

	@property
	def none(self) -> float:
		return self._value

	@property
	def d(self) -> float:
		return self.deci()

	@property
	def c(self) -> float:
		return self.centi()

	@property
	def m(self) -> float:
		return self.mili()

	@property
	def u(self) -> float:
		return self.micro()

	@property
	def n(self) -> float:
		return self.nano()

	@property
	def p(self) -> float:
		return self.pico()

	@property
	def f(self) -> float:
		return self.femto()

	@property
	def a(self) -> float:
		return self.atto()

	@property
	def z(self) -> float:
		return self.zepto()

	@property
	def y(self) -> float:
		return self.yocto()

	@Y.setter
	def Y(self, new_value: float):
		return self.set_yotta(new_value)

	@Z.setter
	def Z(self, new_value: float):
		return self.set_zetta(new_value)

	@E.setter
	def E(self, new_value: float):
		return self.set_exa(new_value)

	@P.setter
	def P(self, new_value: float):
		return self.set_peta(new_value)

	@T.setter
	def T(self, new_value: float):
		return self.set_tera(new_value)

	@G.setter
	def G(self, new_value: float):
		return self.set_giga(new_value)

	@M.setter
	def M(self, new_value: float):
		return self.set_mega(new_value)

	@k.setter
	def k(self, new_value: float):
		return self.set_kilo(new_value)

	@h.setter
	def h(self, new_value: float):
		return self.set_hecto(new_value)

	@da.setter
	def da(self, new_value: float):
		return self.set_deca(new_value)

	@none.setter
	def none(self, new_value: float):
		return self.set_no_prefix(new_value)

	@d.setter
	def d(self, new_value: float):
		return self.set_deci(new_value)

	@c.setter
	def c(self, new_value: float):
		return self.set_centi(new_value)

	@m.setter
	def m(self, new_value: float):
		return self.set_mili(new_value)

	@u.setter
	def u(self, new_value: float):
		return self.set_micro(new_value)

	@n.setter
	def n(self, new_value: float):
		return self.set_nano(new_value)

	@p.setter
	def p(self, new_value: float):
		return self.set_pico(new_value)

	@f.setter
	def f(self, new_value: float):
		return self.set_femto(new_value)

	@a.setter
	def a(self, new_value: float):
		return self.set_atto(new_value)

	@z.setter
	def z(self, new_value: float):
		return self.set_zepto(new_value)

	@y.setter
	def y(self, new_value: float):
		return self.set_yocto(new_value)


pyquantity/commonQuantity.py
"""Module with common quantities"""
from . import Quantity, BaseUnit, DerivedUnit, PrefixEnum, Prefix
from . import errors


class Time(Quantity):
	
	def __init__(self, value: float, prefix: Prefix = PrefixEnum.none):
		Quantity.__init__(self, value, unit=BaseUnit.second(), base_prefix=PrefixEnum.none, prefix=prefix)


class Lenght(Quantity):
	
	def __init__(self, value: float, prefix: Prefix = PrefixEnum.none):
		Quantity.__init__(self, value, unit=BaseUnit.metre(), base_prefix=PrefixEnum.none, prefix=prefix)


class Mass(Quantity):
	
	def __init__(self, value: float, prefix: Prefix = PrefixEnum.kilo):
		Quantity.__init__(self, value, unit=BaseUnit.gram(), base_prefix=PrefixEnum.kilo, prefix=prefix)


class Velocity(Quantity):
	
	def __init__(self, value: float, prefix: Prefix = PrefixEnum.none):
		raise NotImplemented


class AbsoluteTemperature(Quantity):

	ORIGIN_CELCIUS = -273.15
	ORIGIN_FAHRENHEIT = -459.67
	
	def __init__(self, value: float, prefix: Prefix = PrefixEnum.none):
		Quantity.__init__(self, value, unit=BaseUnit.kelvin(), base_prefix=PrefixEnum.none, prefix=prefix)

	@property
	def celcius(self) -> float:
		return self._value + self.ORIGIN_CELCIUS

	@property
	def fahrenheit(self) -> float:
		return self._value + self.ORIGIN_FAHRENHEIT

	@celcius.setter
	def celcius(self, new_value: float):
		self.value = new_value - self.ORIGIN_CELCIUS

	@fahrenheit.setter
	def fahrenheit(self, new_value: float):
		self.value = new_value - self.ORIGIN_FAHRENHEIT

	@classmethod
	def from_celcius(cls, value: float):
		return AbsoluteTemperature(value - self.ORIGIN_CELCIUS)

	@classmethod
	def from_fahrenheit(cls, value: float):
		return AbsoluteTemperature(value - self.ORIGIN_FAHRENHEIT)


class CelsiusTemperature(Quantity):
	
	def __init__(self, value: float):
		Quantity.__init__(self, value, unit=DerivedUnit.celcius(), base_prefix=PrefixEnum.none, prefix=PrefixEnum.none)

	def value_to_prefix(self, prefix: Prefix) -> float:
		raise errors.NoPrefixSupported()

	def set_value_from_prefix(self, new_value: float, prefix: Prefix):
		raise errors.NoPrefixSupported()

	@property
	def kelvin(self) -> float:
		return self._value - AbsoluteTemperature.ORIGIN_CELCIUS

	@property
	def fahrenheit(self) -> float:
		return 1.8 * self._value + 32.

	@kelvin.setter
	def kelvin(self, new_value: float):
		self.value = new_value + AbsoluteTemperature.ORIGIN_CELCIUS

	@fahrenheit.setter
	def fahrenheit(self, new_value: float):
		self.value = 5/9 (new_value - 32.)

	def get_absolute_instance(self) -> AbsoluteTemperature:
		return AbsoluteTemperature(self.kelvin)

	@classmethod
	def from_kelvin(cls, value: float):
		return cls(value + AbsoluteTemperature.ORIGIN_CELCIUS)

	@classmethod
	def from_fahrenheit(cls, value: float):
		return cls(5/9 (value - 32.))

	@classmethod
	def from_absolute_instance(cls, temperature: AbsoluteTemperature):
		if not isinstance(temperature, AbsoluteTemperature):
			raise ValueError(f"'temperature' must be an instance of AbsoluteTemperature, not {type(temperature)}")
		return cls(temperature.celcius)




pyquantity/unit.py
from collections import namedtuple
from typing import Union, List, Tuple, Generator

from . import errors


Prefix = namedtuple('Prefix', ('name', 'symbol', 'value'))


class PrefixEnum:
	yotta = Prefix('yotta', 'Y', 1e24)
	zetta = Prefix('zetta', 'Z', 1e21)
	exa = Prefix('exa', 'E', 1e18)
	peta = Prefix('peta', 'P', 1e15)
	tera = Prefix('tera', 'T', 1e12)
	giga = Prefix('giga', 'G', 1e9)
	mega = Prefix('mega', 'M', 1e6)
	kilo = Prefix('kilo', 'k', 1e3)
	hecto = Prefix('hecto', 'h', 1e2)
	deca = Prefix('deca', 'da', 1e1)
	none = Prefix('', '', 1e0)
	deci = Prefix('deci', 'd', 1e-1)
	centi = Prefix('centi', 'c', 1e-2)
	mili = Prefix('mili', 'm', 1e-3)
	micro = Prefix('micro', 'u', 1e-6)
	nano = Prefix('nano', 'n', 1e-9)
	pico = Prefix('pico', 'p', 1e-12)
	femto = Prefix('femto', 'f', 1e-15)
	atto = Prefix('atto', 'a', 1e-18)
	zepto = Prefix('zepto', 'z', 1e-21)
	yocto = Prefix('yocto', 'y', 1e-24)


	_values_to_names = {
		1e24: yotta,
		1e21: zetta,
		1e18: exa,
		1e15: peta,
		1e12: tera,
		1e9: giga,
		1e6: mega,
		1e3: kilo,
		1e2: hecto,
		1e1: deca,
		1e0: none,
		1e-1: deci,
		1e-2: centi,
		1e-3: mili,
		1e-6: micro,
		1e-9: nano,
		1e-12: pico,
		1e-15: femto,
		1e-18: atto,
		1e-21: zepto,
		1e-24: yocto
	}

	@classmethod
	def add_prefix(cls, name: str, symbol: str, value: float):
		if not isinstance(name, str):
			raise ValueError(f"'name' must be a str not a '{type(name)}'")
		if not isinstance(symbol, str):
			raise ValueError(f"'symbol' must be a str not a '{type(symbol)}'")
		if not isinstance(value, float):
			raise ValueError(f"'value' must be a float not a '{type(value)}'")
		setattr(cls, name, Prefix(name, symbol, value))

	@classmethod
	def remove_prefix(cls, name: str):
		if not isinstance(name, str):
			raise ValueError(f"'name' must be a str not a '{type(name)}'")
		if name in dir(cls):
			return delattr(cls, name)
		raise errors.PrefixError(f"The prefix '{name}' does not exist")

	@classmethod
	def _get_all_prefix(cls) -> List[Prefix]:
		return [attr_ for attr_ in dir(cls) if isinstance(attr_, Prefix)]

	@classmethod
	def _get_all_symbols(cls) -> Generator:
		return (attr_.symbol for attr_ in dir(cls) if isinstance(attr_, Prefix))

	@classmethod
	def _get_prefix_with_value(cls, value: float) -> Prefix:
		try:
			return cls._values_to_names[float(value)]
		except TypeError:
			raise ValueError(f"'value' must be a float, not {type(value)}")
		except KeyError:
			raise errors.PrefixError(f"The prefix with value '{value}' does not exist")

	@classmethod
	def _get_prefix_with_symbol(cls, symbol: str) -> Prefix:
		for prefix in cls._get_all_prefix():
			if prefix.symbol == symbol:
				return prefix
			raise errors.PrefixError(f"The prefix symbol '{symbol}' does not exist")

	@classmethod
	def get_prefix(cls, arg: Union[str, float]) -> Prefix:
		if isinstance(arg, str) and arg in dir(PrefixEnum):
			return getattr(cls, arg)

		elif isinstance(arg, str) and arg in cls._get_all_symbols():
			return _get_prefix_with_symbol(arg)

		elif isinstance(arg, float):
			return cls._get_prefix_with_value(arg)

		elif isinstance(arg, str):
			raise errors.PrefixError(f"The prefix name or symbol '{arg}' does not exist")

		raise ValueError(f"'arg' must be a str or a float, not a {type(arg)}")

	@classmethod
	def get_prefix_with_rank(cls, rank: int):
		"""Return the prefix for the rank given.
		The rank define with: rank = int(log10(value)) // 3.
		If abs(rank) > 8, it's prefix yotta or yocto which is returned.

		Parameters
		----------
		rank : int
			The rank of prefix

		Returns
		-------
		prefix : Prefix
			The prefix corresponding to the rank.
		"""
		try:
			return cls._get_prefix_with_value(pow(10, int(rank)*3))
		except TypeError:
			raise ValueError(f"'rank' must be an integer, not {type(rank)}")
		except errors.PrefixError as e:
			if abs(rank) > 8:
				return cls._get_prefix_with_value(pow(10, (rank / abs(rank)) * 8))
			raise e

	@classmethod
	def convert_value(cls, value: float, to_: Union[str, float, Prefix], from_: Union[str, float, Prefix] = None) -> float:
		try:
			value = float(value)
		except Exception:
			raise ValueError(f"'value' must be a float, not a {type(value)}")

		if isinstance(to_, (str, float)):
			to_ = cls.get_prefix(to_)
		elif not isinstance(to_, Prefix):
			raise ValueError(f"'to_' must be a str, float or a Prefix, not a {type(to_)}")

		if isinstance(from_, (str, float)):
			from_ = cls.get_prefix(from_)
		elif from_ is None:
			from_ = cls.none
		elif not isinstance(from_, Prefix):
			raise ValueError(f"'from_' must be a str, float or a Prefix, not a {type(from_)}")

		return value / (to_.value / from_.value)


class Unit:
	_name: str = ""
	_symbol: str = ""
	_description: str = ""
	_base_unit: bool = False

	def __init__(self, name: str, symbol: str, description: str = ""):
		self.name = name
		self.symbol = symbol
		self.description = description

	def __repr__(self) -> str:
		return f"<Units {self.name} ({self.symbol}). {self.description}>"

	def __str__(self) -> str:
		return f"{self.name} ({self.symbol})"

	@property
	def name(self) -> str:
		return self._name

	@property
	def symbol(self) -> str:
		return self._symbol

	@property
	def description(self) -> str:
		return self._description

	@name.setter
	def name(self, new_name: str):
		if not isinstance(new_name, str):
			raise ValueError(f"'name' must be a str not a {type(new_name)}")
		self._name = new_name
			
	@symbol.setter
	def symbol(self, new_symbol: str):
		if not isinstance(new_symbol, str):
			raise ValueError(f"'symbol' must be a str not a {type(new_name)}")
		self._symbol = new_symbol

	@description.setter
	def description(self, new_desc: str):
		if not isinstance(new_desc, str):
			raise ValueError(f"'description' must be a str not a {type(new_desc)}")
		self._description = new_desc

	def is_base_unit(self) -> bool:
		return self._base_unit

	def is_derived_unit(self) -> bool:
		return not self._base_unit

	def name_with_prefix(self, prefix: Union[str, float, Prefix]) -> str:
		if isinstance(prefix, Prefix):
			return prefix.name + self.name
		elif isinstance(prefix, (str, float)):
			return PrefixEnum.get_prefix(prefix).name + self.name
		raise ValueError(f"'prefix' must be a str, a float or a Prefix, not {type(prefix)}")

	def symbol_with_prefix(self, prefix: Union[str, float, Prefix]) -> str:
		if isinstance(prefix, Prefix):
			return prefix.symbol + self.symbol
		elif isinstance(prefix, (str, float)):
			return PrefixEnum.get_prefix(prefix).symbol + self.symbol
		raise ValueError(f"'prefix' must be a str, a float or a Prefix, not {type(prefix)}")


class BaseUnit(Unit):
	_base_unit = True

	@staticmethod
	def metre():
		return BASE_UNITS['metre']

	@staticmethod
	def meter():
		return BASE_UNITS['meter']

	@staticmethod
	def second():
		return BASE_UNITS['second']

	@staticmethod
	def gram():
		return BASE_UNITS['gram']

	@staticmethod
	def kelvin():
		return BASE_UNITS['kelvin']

	@staticmethod
	def mole():
		return BASE_UNITS['mole']

	@staticmethod
	def ampere():
		return BASE_UNITS['ampere']

	@staticmethod
	def candela():
		return BASE_UNITS['candela']

	@staticmethod
	def radian():
		return BASE_UNITS['radian']

	@staticmethod
	def steradian():
		return BASE_UNITS['steradian']


class DerivedUnit(Unit):
	
	@staticmethod
	def celcius():
		return DERIVED_UNIT['celcius']

	@staticmethod
	def fahrenheit():
		return DERIVED_UNIT['fahrenheit']


BASE_UNITS = {
	'metre': BaseUnit('Metre', 'm', description="The base unit of lenght."),
	'meter': BaseUnit('Meter', 'm', description="The base unit of lenght"),
	'second': BaseUnit('Second', 's', description="The base unit of time"),
	'gram': BaseUnit('Gram', 'g', description="The base unit of mass"),
	'kelvin': BaseUnit('Kelvin', 'K', description="The base unit of temperature"),
	'mole': BaseUnit('Mole', 'mol', description="The base unit of amount of substance"),
	'ampere': BaseUnit('Ampere', 'A', description="The base unit of electric current"),
	'candela': BaseUnit('Candela', 'cd', description="The base unit of luminous intensity"),
	'radian': BaseUnit('Radian', 'rad', description="The base unit of plane angle"),
	'steradian': BaseUnit('Steradian', 'sr', description="The base unit of solid angle")
}


DERIVED_UNIT = {
	'celcius': DerivedUnit('Celcius', '°C', description="The unit of Celcius temperature"),
	'fahrenheit': DerivedUnit('Fahrenheit', '°F', description="The unit of Fahrenheit temperature")
}